doc/config_reload/config_reload_enhancement.md

```diff
@@ -0,0 +1,122 @@
+- Services can be configured to be started immediately or delayed. This can be using a field in FEATURE table.
+- The existing timers should be removed by this event driven approach.
+- This flow is applicable in case of all reboots (warm/fast/cold) as well as config reload.
+
+
+### High-Level Design
+Currently hostcfgd controls the services based on the feature table. The feature table has a specific field 'has_timer' for the non essential services which needs to be delayed during the reboot flow. This field will be now replaced by new field called "delayed". These services will controlled by hostcfgd.
+During the hostcfgd initialization it will cache these delayed services based on the configuration in the feature table. The hostcfgd will also subscribe to PORT_TABLE in the APPL_DB. Once the switch is initialized and all the ports are created in ASIC and Kernel the PortSyncd will publish PortInitDone key in the APPL_DB. On receiving this key the hostcfgd will go through the delayed services list and enables them. There will also be a timeout defined for the services to start if PortInitDone is not defined within the specific timeout. This is to ensure the management related services start even if there is some failure for the switch to initialize.
+
+
+The below diagram explains the sequence when config reload is executed. 
+![](/images/config_reload/Enhance_config_reload.JPG)
+The below diagram explains the new sequence inside hostcfgd.
+![](/images/config_reload/Hostcfgd_flow_for_delayed_services.JPG)
+### SAI API Requirements
+None
+
+### Configuration and management
+
+#### Config command
+
+No new commands are introduced as part of this design.
+
+#### Show command
+
+No new commands are introduced as part of this design
+
+#### DB Migrator
+
+The 'has_timer' field in FEATURE table will be changed to 'delayed'. Hence db_migrator is required to modify the configurations to reflect this change.
+
+### YANG model changes
+
+Yang model needs to be updated for FEATURE_TABLE. The 'has_timer' field will be removed and replaced with 'delayed'
+
+```
+                leaf delayed {
+                    description "This configuration identicates if the feature needs to be delayed until
+                                 system initialization";
+                    type stypes:boolean_type;
+                    default "false";
+                }
+
+```
+
+### Warmboot and Fastboot Considerations
+
+In case of Warmboot and fastboot, Hostcfgd itself is currently delayed today by a timer. This timer will also be removed and replaced by hostcfgd waiting for warmboot/fastboot completion using waitAdvancedBootDone.
+Additionally there will be checks added to Warmboot command to ensure all services including delayed services are started before proceeding.
+
+### Testing Design
+
+#### Unit tests
+Hostcfgd tests would be enhanced to cover the new flow.
+- Ensure the service start are done for critical services immediately without delay.

```

Comment:
What is the criteria to decide whether a service is critical or not?

Reply:
It is determined as per the existing approach where we already delay a list of services. In my opinion all the services that are required for faster initialization and traffic convergence are critical. All monitoring services like snmp, lldp, telemetry can be delayed